File: refactor.py
import re
import os

with open('cmd/crompressor/main.go', 'r') as f:
    original_code = f.read()

funcs_to_move = [
    'daemonCmd', 'sharedDaemonCmd', 'keysCmd', 'trustCmd', 'shareCmd', 
    'mountCmd', 'cromfsCmd', 'llmVfsCmd'
]

sys_code = """//go:build !wasm
package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/MrJc01/crompressor/internal/autobrain"
	"github.com/MrJc01/crompressor/internal/metrics"
	"github.com/MrJc01/crompressor/internal/network"
	"github.com/MrJc01/crompressor/internal/vfs"
	cvfs "github.com/MrJc01/crompressor/pkg/cromlib/vfs"
	"github.com/MrJc01/crompressor/internal/cromfs"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/spf13/cobra"
)

func addSystemCommands(rootCmd *cobra.Command) {
	rootCmd.AddCommand(daemonCmd())
	rootCmd.AddCommand(sharedDaemonCmd())
	rootCmd.AddCommand(shareCmd())
	rootCmd.AddCommand(keysCmd())
	rootCmd.AddCommand(trustCmd())
	rootCmd.AddCommand(mountCmd())
	rootCmd.AddCommand(cromfsCmd())
	rootCmd.AddCommand(llmVfsCmd())
}
"""

new_main = original_code

for fn in funcs_to_move:
    # Pattern to match the whole function. It assumes the function ends with "\n}\n"
    # and has no nested "\n}\n" inside. This is usually safe enough if the file is well-formatted.
    pattern = r"func " + fn + r"\(\) \*cobra\.Command \{.*?\n\}\n"
    match = re.search(pattern, new_main, re.DOTALL)
    if match:
        sys_code += "\n" + match.group(0)
        new_main = new_main.replace(match.group(0), "")
    else:
        print(f"Function {fn} not found!")

# Remove imports from main.go that are only used in sys_commands
imports_to_remove = [
    '"github.com/MrJc01/crompressor/internal/network"',
    '"github.com/MrJc01/crompressor/internal/vfs"',
    'cvfs "github.com/MrJc01/crompressor/pkg/cromlib/vfs"',
    '"github.com/MrJc01/crompressor/internal/cromfs"',
    '"github.com/prometheus/client_golang/prometheus/promhttp"',
]

for imp in imports_to_remove:
    new_main = new_main.replace("\t" + imp + "\n", "")

with open('cmd/crompressor/sys_commands.go', 'w') as f:
    f.write(sys_code)

with open('cmd/crompressor/main.go', 'w') as f:
    f.write(new_main)

print("Done factoring main.go.")

File: refactor.go
package main

import (
	"bytes"
	"go/ast"
	"go/format"
	"go/parser"
	"go/token"
	"os"
)

func main() {
	fset := token.NewFileSet()
	f, err := parser.ParseFile(fset, "cmd/crompressor/main.go", nil, parser.ParseComments)
	if err != nil {
		panic(err)
	}

	funcsToMove := map[string]bool{
		"daemonCmd":       true,
		"sharedDaemonCmd": true,
		"keysCmd":         true,
		"trustCmd":        true,
		"shareCmd":        true,
		"mountCmd":        true,
		"cromfsCmd":       true,
		"llmVfsCmd":       true,
	}

	var sysFuncs []ast.Decl
	var mainFuncs []ast.Decl

	for _, decl := range f.Decls {
		if fn, ok := decl.(*ast.FuncDecl); ok {
			if funcsToMove[fn.Name.Name] {
				sysFuncs = append(sysFuncs, decl)
				continue
			}
		}
		mainFuncs = append(mainFuncs, decl)
	}

	// Remove imports
	for _, decl := range mainFuncs {
		if genDecl, ok := decl.(*ast.GenDecl); ok && genDecl.Tok == token.IMPORT {
			var newSpecs []ast.Spec
			for _, spec := range genDecl.Specs {
				if impSpec, ok := spec.(*ast.ImportSpec); ok {
					path := impSpec.Path.Value
					if path == `"github.com/MrJc01/crompressor/internal/network"` ||
						path == `"github.com/MrJc01/crompressor/internal/vfs"` ||
						path == `"github.com/MrJc01/crompressor/pkg/cromlib/vfs"` ||
						path == `"github.com/MrJc01/crompressor/internal/cromfs"` ||
						path == `"github.com/prometheus/client_golang/prometheus/promhttp"` {
						continue
					}
				}
				newSpecs = append(newSpecs, spec)
			}
			genDecl.Specs = newSpecs
		}
	}

	// Write new main.go
	f.Decls = mainFuncs
	var buf bytes.Buffer
	format.Node(&buf, fset, f)
	os.WriteFile("cmd/crompressor/main.go", buf.Bytes(), 0644)

	// Write sys_commands.go
	sysTokens := token.NewFileSet()
	sysF := &ast.File{
		Name: &ast.Ident{Name: "main"},
		Decls: sysFuncs,
	}
	var sysBuf bytes.Buffer
	sysBuf.WriteString(`//go:build !wasm

package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/MrJc01/crompressor/internal/autobrain"
	"github.com/MrJc01/crompressor/internal/metrics"
	"github.com/MrJc01/crompressor/internal/network"
	"github.com/MrJc01/crompressor/internal/vfs"
	cvfs "github.com/MrJc01/crompressor/pkg/cromlib/vfs"
	"github.com/MrJc01/crompressor/internal/cromfs"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/spf13/cobra"
)

func addSystemCommands(rootCmd *cobra.Command) {
	rootCmd.AddCommand(daemonCmd())
	rootCmd.AddCommand(sharedDaemonCmd())
	rootCmd.AddCommand(shareCmd())
	rootCmd.AddCommand(keysCmd())
	rootCmd.AddCommand(trustCmd())
	rootCmd.AddCommand(mountCmd())
	rootCmd.AddCommand(cromfsCmd())
	rootCmd.AddCommand(llmVfsCmd())
}

`)
	format.Node(&sysBuf, sysTokens, sysF)
	os.WriteFile("cmd/crompressor/sys_commands.go", sysBuf.Bytes(), 0644)
}
